{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import feather\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import numpy as np\n",
    "import copy\n",
    "import openpyxl\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_ncounts_ngenes(ad): \n",
    "    # add counts per well\n",
    "    ad.obs['n_counts'] = ad.X.sum(axis=1)\n",
    "\n",
    "    # add genes per well\n",
    "    ad.obs['n_genes'] = ad.X.astype('bool').sum(axis=1)\n",
    "    \n",
    "def import_feather(path, seq_type, sample_batch):\n",
    "    \"\"\"\n",
    "    Converts feather objects to useable anndata. Ignores data feather objects other than $readcount$exon$all.\n",
    "\n",
    "    Args:\n",
    "        path: Path to directory with all feather objects\n",
    "        \n",
    "    Returns:\n",
    "        Anndata object containing raw read counts per gene\n",
    "    \"\"\"\n",
    "    \n",
    "    DATA_DIR = 'FFPE/seq_data/'\n",
    "    \n",
    "    FILE_ensembl_to_gene = DATA_DIR+path+\"/ensemblID_to_gene.feather\"\n",
    "    FILE_matrix_genes = DATA_DIR+path+\"/matrix_genes.feather\"\n",
    "    FILE_counts = DATA_DIR+path+\"/zUMI_dataframe_counts.feather\"\n",
    "    FILE_umi = DATA_DIR+path+\"/zUMI_dataframe_umi.feather\"\n",
    "    FILE_sample_sheet = DATA_DIR+path+\"/sample_sheet.txt\"\n",
    "\n",
    "    # load sample sheet\n",
    "    sample_sheet = pd.read_csv(FILE_sample_sheet, sep=\"\\t\")\n",
    "    if 'PCRcycles' in sample_sheet.columns: \n",
    "        sample_sheet['PCR_cycles'] = sample_sheet['PCRcycles'].copy()\n",
    "\n",
    "    # load ensembl ID to gene name conversion \n",
    "    gene_id_to_name_DF = feather.read_dataframe(FILE_ensembl_to_gene)\n",
    "    gene_name_dict = pd.Series(gene_id_to_name_DF.gene_name.values,index=gene_id_to_name_DF.gene_id).to_dict()\n",
    "\n",
    "    # load matrix genes & replace ensembl ID with gene name\n",
    "    matrix_genes_DF = feather.read_dataframe(FILE_matrix_genes)\n",
    "    matrix_genes_DF['gene_name'] = ''\n",
    "    for idx, row in matrix_genes_DF.iterrows(): \n",
    "        matrix_genes_DF.at[idx,'gene_name'] = gene_name_dict[row[0]]\n",
    "\n",
    "    # load barcode to well position conversion \n",
    "    if 'sample' not in sample_sheet.columns: \n",
    "        if 'Sample' in sample_sheet.columns: \n",
    "            sample_sheet['sample'] = sample_sheet['Sample'].copy()\n",
    "            \n",
    "    bc_to_well = dict(zip(sample_sheet[\"concat_bc\"], sample_sheet[\"sample\"]))\n",
    "\n",
    "    # load raw read count dataframe (if UMIs available, use it)\n",
    "    if os.path.isfile(FILE_umi): \n",
    "        DF = feather.read_dataframe(FILE_umi)\n",
    "    else: \n",
    "        DF = feather.read_dataframe(FILE_counts)\n",
    "        \n",
    "    DF['gene_name'] = matrix_genes_DF['gene_name']  # add gene name column\n",
    "    DF.set_index('gene_name',inplace=True)  # set gene name as index\n",
    "    DF.rename(columns=bc_to_well,inplace=True) # rename columns from barcodes to well positions\n",
    "    umi_cts = DF.transpose()\n",
    "    umi_cts.fillna(0,inplace = True)\n",
    "\n",
    "    # build obs dataframe with sample attribute values\n",
    "    obs_df = sample_sheet.set_index('sample')\n",
    "    obs_df = obs_df[[x for x in sample_sheet.columns if x not in [\"sample\",\"concat_bc\"]]]\n",
    "    tokeep = [x for x in umi_cts.index]\n",
    "    obs_df = obs_df.loc[tokeep]\n",
    "    obs_df[\"seq_type\"] = seq_type\n",
    "    obs_df[\"sample_batch\"] = sample_batch\n",
    "\n",
    "    # build var dataframe\n",
    "    var_df = pd.DataFrame([],index=umi_cts.columns)\n",
    "\n",
    "    # build anndata object\n",
    "    cts_ad = anndata.AnnData(X=umi_cts.values, obs=obs_df, var = var_df)\n",
    "    cts_ad.var_names_make_unique()\n",
    "    \n",
    "    add_ncounts_ngenes(cts_ad)\n",
    "\n",
    "    print(cts_ad.obs.head(1))\n",
    "    print(cts_ad.shape)\n",
    "    \n",
    "    return cts_ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_hg19_and_mm10(ad):\n",
    "    def calc_cts(ad, cts_type):\n",
    "        ad.obs['n_counts_%s' %cts_type] = ad.X.sum(axis=1)\n",
    "        return ad\n",
    "    def calc_genes(ad, gene_type): \n",
    "        ad.obs['n_genes_%s' %gene_type] = ad.X.astype(bool).sum(axis=1)\n",
    "        return ad\n",
    "    \n",
    "    hg19_genes = [g for g in ad.var_names if g.startswith('hg19_')]\n",
    "    hg19_gdict = dict(zip(hg19_genes,[g.lstrip('hg19') for g in hg19_genes]))\n",
    "    for key, item in hg19_gdict.items(): \n",
    "        hg19_gdict[key] = item[1:]\n",
    "        \n",
    "    mm10_genes = [g for g in ad.var_names if g.startswith('mm10_')]\n",
    "    mm10_gdict = dict(zip(mm10_genes,[g.lstrip('mm10') for g in mm10_genes]))\n",
    "    for key, item in mm10_gdict.items(): \n",
    "        mm10_gdict[key] = item[1:]\n",
    "    \n",
    "    ad_hg19 = ad[:,hg19_genes]\n",
    "    ad_hg19.var.rename(index=hg19_gdict, inplace=True)\n",
    "    ad_hg19.var_names = ad_hg19.var.index\n",
    "    print(ad_hg19.shape)\n",
    "    \n",
    "    ad_mm10 = ad[:,mm10_genes]\n",
    "    ad_mm10.var.rename(index=mm10_gdict, inplace=True)\n",
    "    ad_mm10.var_names = ad_mm10.var.index\n",
    "    print(ad_mm10.shape)\n",
    "    \n",
    "    calc_cts(ad_hg19, 'hg19')\n",
    "    calc_genes(ad_hg19, 'hg19')\n",
    "    calc_cts(ad_mm10, 'mm10')\n",
    "    calc_genes(ad_mm10, 'mm10')\n",
    "    \n",
    "    ad_all = ad.copy()\n",
    "    ad_all.obs['n_counts_hg19'] = ad_hg19.obs['n_counts_hg19']\n",
    "    ad_all.obs['n_genes_hg19'] = ad_hg19.obs['n_genes_hg19']\n",
    "    ad_all.obs['n_counts_mm10'] = ad_mm10.obs['n_counts_mm10']\n",
    "    ad_all.obs['n_genes_mm10'] = ad_mm10.obs['n_genes_mm10']\n",
    "    \n",
    "    ad_all.obs['percent_mm10'] = 100*np.sum(ad_all[:, mm10_genes].X, axis=1) / np.sum(ad_all.X, axis=1)\n",
    "    ad_all.obs['percent_hg19'] = 100*np.sum(ad_all[:, hg19_genes].X, axis=1) / np.sum(ad_all.X, axis=1)\n",
    "    \n",
    "    ad_mm10.obs['percent_mm10'] = ad_all.obs['percent_mm10']\n",
    "    ad_mm10.obs['percent_hg19'] = ad_all.obs['percent_hg19']\n",
    "    \n",
    "    print('Returned ad_hg19, ad_mm10, ad_all')\n",
    "    return ad_hg19, ad_mm10, ad_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_adata_to_csvs(ad, filename, writemode='default'):\n",
    "    import pandas as pd\n",
    "    \n",
    "    out_dir = 'write/submission_files'\n",
    "    \n",
    "    # gene expression matrix\n",
    "    if writemode=='default':\n",
    "        df = pd.DataFrame(data=ad.X, \n",
    "                     index=ad.obs_names, \n",
    "                     columns=ad.var_names)\n",
    "    elif writemode=='counts':\n",
    "        df = pd.DataFrame(data=ad.layers['counts'].toarray(),\n",
    "                     index=ad.obs_names, \n",
    "                     columns=ad.var_names)\n",
    "    df.T.to_csv('%s/%s_matrix.csv' %(out_dir, filename))\n",
    "\n",
    "    # gene names\n",
    "    ad.var.to_csv('%s/%s_var.csv' %(out_dir, filename))\n",
    "\n",
    "    # barcodes / nuclei and metadata\n",
    "    ad.obs.to_csv('%s/%s_obs.csv' %(out_dir, filename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_mito_ncounts(ad, case='mouse'):\n",
    "    # calculate mitochondrial fraction\n",
    "    if case=='human': \n",
    "        mito_genes = ad.var_names.str.startswith('MT-') # human\n",
    "    if case=='hg19': \n",
    "        mito_genes = ad.var_names.str.startswith('hg19_MT-') # human\n",
    "    elif case=='mouse': \n",
    "        mito_genes = ad.var_names.str.startswith('mt-') #mouse\n",
    "    \n",
    "    ad.obs['frac_mito'] = np.sum(ad[:, mito_genes].X, axis=1) / np.sum(ad.X, axis=1)\n",
    "\n",
    "    # add the total counts per cell as observations-annotation to adata\n",
    "    ad.obs['n_counts'] = ad.X.sum(axis=1)\n",
    "    return ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_batch_logical_gate(ad, blims, feat='n_genes'):\n",
    "    import operator\n",
    "#     batches = list(set(ad.obs['batch']))\n",
    "    batches = ad.obs['batch'].cat.categories\n",
    "    combined_bool = ad.obs['batch']==None\n",
    "    for batch in batches: \n",
    "        lims_b = blims[batch]\n",
    "        bool_b = (ad.obs['batch']==batch) & \\\n",
    "                    (ad.obs[feat]>lims_b[0]) & \\\n",
    "                    (ad.obs[feat]<lims_b[1])\n",
    "        combined_bool = operator.or_(combined_bool, bool_b)\n",
    "    return combined_bool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cells_per_gene(ad): \n",
    "    return ad.X.toarray().astype(bool).sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniform_jitter(obj_len, minval=-2, maxval=2): \n",
    "    return np.random.uniform(minval,maxval,obj_len)\n",
    "\n",
    "def scatter_hg19_vs_mm10(ad, hue_var='plate', savefig=False): \n",
    "    colors = sns.color_palette()\n",
    "    fig, ax = plt.subplots(figsize=(5,5))\n",
    "    g = sns.scatterplot(x=ad.obs['percent_hg19'] + uniform_jitter(ad.obs.shape[0]),\n",
    "                y=ad.obs['percent_mm10'] + uniform_jitter(ad.obs.shape[0]),\n",
    "                data=ad.obs,\n",
    "                hue=ad.obs[hue_var],\n",
    "                s=12,\n",
    "                edgecolor='k',\n",
    "                linewidth=0.3,\n",
    "                alpha=0.5)\n",
    "    g.legend(loc='upper right', ncol=1, bbox_to_anchor=(1.5, 0.5))\n",
    "    plt.xlim([-10,110])\n",
    "    plt.ylim([-10,110])\n",
    "    plt.xlabel('% hg19 gene counts',fontsize=16)\n",
    "    plt.ylabel('% mm10 gene counts',fontsize=16)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "    if savefig: \n",
    "        plt.savefig('%s/scatter_mm10_hg19.pdf' %(sc.settings.figdir), bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scatter_UMI_genes_hist(ad, samplestr, density=True, savefig=True): \n",
    "    from scipy.stats import gaussian_kde\n",
    "\n",
    "    # definitions for the axes\n",
    "    left, width = 0.1, 0.65\n",
    "    bottom, height = 0.1, 0.65\n",
    "    spacing = 0.005\n",
    "\n",
    "    rect_scatter = [left, bottom, width, height]\n",
    "    rect_histx = [left, bottom + height + spacing, width, 0.2]\n",
    "    rect_histy = [left + width + spacing, bottom, 0.2, height]\n",
    "\n",
    "    h = plt.figure(figsize=(3,3), dpi=1200)\n",
    "\n",
    "    ax_scatter = plt.axes(rect_scatter)\n",
    "    ax_scatter.tick_params(direction='in', top=True, right=True)\n",
    "    ax_histx = plt.axes(rect_histx)\n",
    "    ax_histx.tick_params(direction='in', labelbottom=False)\n",
    "    ax_histy = plt.axes(rect_histy)\n",
    "    ax_histy.tick_params(direction='in', labelleft=False)\n",
    "\n",
    "    # plot x=y line\n",
    "    ax_scatter.plot([0,6],[0,6],'k--',linewidth=1)\n",
    "    \n",
    "    x = np.log10(ad.obs['n_counts'])\n",
    "    y = np.log10(ad.obs['n_genes'])\n",
    "    \n",
    "    # calculate density\n",
    "    if density: \n",
    "        xy = np.vstack([x,y])\n",
    "        z = gaussian_kde(xy)(xy)\n",
    "        ax_sc = ax_scatter.scatter(x, y, c=z, s=3, alpha=0.5, cmap='coolwarm')\n",
    "        plt.colorbar(ax_sc)\n",
    "    \n",
    "    else: \n",
    "\n",
    "        ax_scatter.scatter(x, y, s=3, color='#696969', alpha=0.5)\n",
    "    \n",
    "    bins = np.arange(0,6,0.05)\n",
    "    ax_histx.hist(x, bins=bins, facecolor='#696969')\n",
    "    ax_histy.hist(y, bins=bins, orientation='horizontal', facecolor='#696969')\n",
    "\n",
    "    # set axis properties \n",
    "    ax_scatter.set_xlim([1,5])\n",
    "    ax_scatter.set_xticks([1,2,3,4,5])\n",
    "    ax_scatter.set_ylim([1,4])\n",
    "    ax_scatter.set_yticks([1,2,3,4])\n",
    "    ax_scatter.set_xlabel('log10(n_UMIs)', fontsize=12)\n",
    "    ax_scatter.set_ylabel('log10(n_genes)', fontsize=12)\n",
    "    \n",
    "    ax_histx.set_xlim(ax_scatter.get_xlim())\n",
    "    ax_histy.set_ylim(ax_scatter.get_ylim())\n",
    "    \n",
    "    ax_histx.spines['top'].set_visible(False)\n",
    "    ax_histx.spines['right'].set_visible(False)\n",
    "    ax_histx.spines['left'].set_visible(False)\n",
    "    ax_histx.set_xticks([])\n",
    "    ax_histx.set_yticks([])\n",
    "    \n",
    "    ax_histy.spines['top'].set_visible(False)\n",
    "    ax_histy.spines['right'].set_visible(False)\n",
    "    ax_histy.spines['bottom'].set_visible(False)\n",
    "    ax_histy.set_xticks([])\n",
    "    ax_histy.set_yticks([])\n",
    "\n",
    "    if savefig: \n",
    "        plt.savefig('%s/scatter_ngenes_UMIs_hist_%s.png' %(sc.settings.figdir, samplestr), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_gray_monoscale_cmap(): \n",
    "    import matplotlib\n",
    "    from matplotlib import cm\n",
    "    blues = cm.get_cmap('Blues', 200)\n",
    "    blues_array = blues(np.linspace(0, 1, 15)).tolist()\n",
    "    blues_array.insert(0, [0.85, 0.85, 0.85, 1.0])\n",
    "    bg = matplotlib.colors.ListedColormap(blues_array,name='blues_with_gray')\n",
    "    return bg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_pyannotable(): \n",
    "    import pickle5 as pickle\n",
    "    with open('pyannotables.pickle', 'rb') as handle:\n",
    "        tables = pickle.load(handle)\n",
    "    return tables\n",
    "\n",
    "def load_gene_len_dict(species):\n",
    "    import pickle5 as pickle\n",
    "    \n",
    "    if species=='mouse':\n",
    "        table_type = 'mus_musculus-GRCm38-ensembl100'\n",
    "        dict_name = 'mouse_gene_lengths.pickle'\n",
    "    elif species=='human':\n",
    "        table_type = 'homo_sapiens-GRCh38-ensembl100'\n",
    "        dict_name = 'human_gene_lengths.pickle'\n",
    "    \n",
    "    if os.path.isfile(dict_name): \n",
    "        with open(dict_name, 'rb') as handle:\n",
    "            gene_len_dict = pickle.load(handle)\n",
    "        \n",
    "    else: \n",
    "        tables = load_pyannotable()\n",
    "        gene_len_dict = {}\n",
    "        table = tables[table_type]\n",
    "        genes = tables[table_type].gene_name.unique()\n",
    "        for g in genes: \n",
    "            hits = table[table['gene_name']==g]\n",
    "            if hits.shape[0]>1: \n",
    "                gene_lens = list(set(table[table['gene_name']==g]['gene_coding_length']))\n",
    "                if len(gene_lens)>1:\n",
    "                    gene_len = max(gene_lens)\n",
    "                else: \n",
    "                    gene_len = gene_lens[0]\n",
    "            else: \n",
    "                gene_len = hits['gene_coding_length'].item()\n",
    "            gene_len_dict[g] = gene_len\n",
    "            \n",
    "        # save\n",
    "        with open(dict_name, 'wb') as handle:\n",
    "            pickle.dump(gene_len_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "            \n",
    "    genes_df = pd.DataFrame(gene_len_dict.items(), columns=['Gene','Length'])\n",
    "    genes_df.set_index('Gene', inplace=True)\n",
    "\n",
    "    return genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pathway_genes(pathway_type=['c2']): \n",
    "    pathway_genes = {}\n",
    "    pathways = []\n",
    "    \n",
    "    for path in pathway_type: \n",
    "        fh = open('%s_pathways.txt' %path)\n",
    "        Lines = fh.readlines()\n",
    "        \n",
    "        for line in Lines: \n",
    "            a = line.strip().split(',')\n",
    "            pathway = a[0]\n",
    "            pathways.append(pathway)\n",
    "\n",
    "            genes = a[1:]\n",
    "            pathway_genes[pathway] = genes\n",
    "            \n",
    "        fh.close()\n",
    "    print('Loaded %i pathways' %(len(pathways)))\n",
    "    return pathway_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def program_counts(ad, genes): \n",
    "    vals = ad[:,genes].layers['counts'].toarray().sum(axis=1)\n",
    "    vals_norm = vals/len(genes)\n",
    "    return vals_norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def weighted_program_counts(ad, genes, gene_weights): \n",
    "    cts = ad[:,genes].layers['counts'].toarray()\n",
    "    weighted_sum = [(c_i/gene_weights).sum() for c_i in cts]\n",
    "    return weighted_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_pathway_ct(pathway_dict): \n",
    "    reverse_dict = {}\n",
    "    for k,v in sorted(pathway_dict.items()):\n",
    "        for x in v:\n",
    "            reverse_dict.setdefault(x,[]).append(k)\n",
    "    gene_pathway_ct = {k:len(v) for k,v in sorted(reverse_dict.items())}\n",
    "    \n",
    "    return gene_pathway_ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def weighted_pathway_aggregation(ad, pathways, pathway_dict, gene_pathway_ct): \n",
    "    import numpy as np\n",
    "    import anndata\n",
    "    \n",
    "    pathway_scores = np.empty((ad.shape[0], len(pathways)))\n",
    "    for idx, pway in zip(range(len(pathways)), pathways): \n",
    "        p_genes = pathway_dict[pway]\n",
    "        gene_weights = [gene_pathway_ct[g] for g in p_genes]\n",
    "        pathway_scores[:,idx] = weighted_program_counts(ad, p_genes, gene_weights)\n",
    "    \n",
    "    var_df = pd.DataFrame(pathways, columns=['pathway'])\n",
    "    var_df.set_index('pathway', inplace=True)\n",
    "    ad_path = anndata.AnnData(X=pathway_scores,\n",
    "                                obs=ad.obs[['Sample',\n",
    "                                             'sample_batch',\n",
    "                                             'n_counts_hg19',\n",
    "                                             'n_genes_hg19',\n",
    "                                             'leiden',\n",
    "                                             'plate']],\n",
    "                               var=var_df) \n",
    "    \n",
    "    return ad_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pathways_anndata(ad, pathway_genes, MIN_FRAC=0.75, MIN_GENES=10, MAX_GENES=200, weighted=False):\n",
    "    import anndata\n",
    "    import pandas as pd\n",
    "    \n",
    "    ad_pathways = ad.copy()\n",
    "    valid_pathways = []\n",
    "    valid_genes = []\n",
    "    n_pathway_components = []\n",
    "    updated_dict = {}\n",
    "\n",
    "    print('Keeping only valid genes and pathways')\n",
    "    for path in sorted(pathway_genes.keys()): \n",
    "        path = path.strip()\n",
    "        valid_genes = [g for g in pathway_genes[path] if g in ad_pathways.var_names]\n",
    "\n",
    "        n_genes = len(valid_genes)\n",
    "        if (n_genes > MIN_GENES) & (n_genes < MAX_GENES):\n",
    "            if float(n_genes)/len(pathway_genes[path]) > MIN_FRAC: \n",
    "                if path not in valid_pathways: # remove duplicate pathway names\n",
    "                    valid_pathways.append(path)\n",
    "                    n_pathway_components.append(len(valid_genes))\n",
    "\n",
    "                    # calculate program score (raw counts)\n",
    "                    ad_pathways.obs['%s' %path] = program_counts(ad_pathways, valid_genes)\n",
    "\n",
    "                    updated_dict[path] = valid_genes\n",
    "    \n",
    "    print(len(valid_pathways))\n",
    "#     valid_pathways = list(set(valid_pathways))\n",
    "    \n",
    "    # make new anndata object\n",
    "    if weighted: \n",
    "        print('Making *weighted* anndata object')\n",
    "        gene_pathway_ct = get_gene_pathway_ct(updated_dict)\n",
    "        ad_path = weighted_pathway_aggregation(ad_pathways, \n",
    "                                            valid_pathways, \n",
    "                                            updated_dict, \n",
    "                                            gene_pathway_ct)\n",
    "        \n",
    "    else: \n",
    "        print('Making unweighted anndata object')\n",
    "        pathway_scores = ad_pathways.obs[valid_pathways]\n",
    "        var_df = pd.DataFrame(valid_pathways, columns=['pathway'])\n",
    "        var_df.set_index('pathway', inplace=True)\n",
    "        ad_path = anndata.AnnData(X=pathway_scores.to_numpy(),\n",
    "                        obs=ad_pathways.obs[['Sample','sample_batch',\n",
    "                                             'batch','plate',\n",
    "                                            'n_counts','n_genes',\n",
    "                                            'n_counts_hg19','n_genes_hg19']],\n",
    "                       var=var_df) \n",
    "    \n",
    "    return ad_path, updated_dict, valid_pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genes_heatmap(ad, glist, savefig=False, savestr='', max_scale=5, pd_colors=None):\n",
    "    sns.reset_orig()\n",
    "    glist = [g for g in glist if g in ad.var_names]\n",
    "    ad1 = ad[:,glist].copy()\n",
    "    sc.pp.normalize_per_cell(ad1, counts_per_cell_after=1e4)\n",
    "    sc.pp.log1p(ad1)\n",
    "    sc.pp.scale(ad1)\n",
    "\n",
    "    cts = pd.DataFrame(ad1.X, index=ad1.obs.index)\n",
    "    plt.figure()\n",
    "    cg = sns.clustermap(cts.T, \n",
    "               xticklabels=[],\n",
    "               yticklabels=ad1.var_names,\n",
    "               col_colors=pd_colors,\n",
    "               vmin=-max_scale, vmax=max_scale,\n",
    "              cmap='RdBu_r')\n",
    "    cg.ax_row_dendrogram.set_visible(False)\n",
    "    cg.ax_col_dendrogram.set_visible(False)\n",
    "    cg.ax_heatmap.set_xlabel('Nuclei')\n",
    "    cg.ax_heatmap.set_ylabel('Marker genes')\n",
    "    \n",
    "    if savefig: \n",
    "        plt.savefig('%s/clustermap_%s.pdf' %(sc.settings.figdir, savestr))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cluster_proportions(adata,\n",
    "                            cluster_key=\"leiden\",\n",
    "                            sample_key=\"sample_type\",\n",
    "                            drop_values=None):\n",
    "    \"\"\"\n",
    "    Input\n",
    "    =====\n",
    "    adata : AnnData object\n",
    "    cluster_key : key of `adata.obs` storing cluster info\n",
    "    sample_key : key of `adata.obs` storing sample/replicate info\n",
    "    drop_values : list/iterable of possible values of `sample_key` that you don't want\n",
    "    \n",
    "    Returns\n",
    "    =======\n",
    "    pd.DataFrame with samples as the index and clusters as the columns and 0-100 floats\n",
    "    as values\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_tmp = adata.copy()\n",
    "    sizes = adata_tmp.obs.groupby([cluster_key, sample_key]).size()\n",
    "    props = sizes.groupby(level=1).apply(lambda x: 100 * x / x.sum()).reset_index() \n",
    "    props = props.pivot(columns=sample_key, index=cluster_key).T\n",
    "    props.index = props.index.droplevel(0)\n",
    "    props.fillna(0, inplace=True)\n",
    "    \n",
    "    if drop_values is not None:\n",
    "        for drop_value in drop_values:\n",
    "            props.drop(drop_value, axis=0, inplace=True)\n",
    "    return props\n",
    "\n",
    "\n",
    "def plot_cluster_proportions(cluster_props, \n",
    "                             cluster_palette=None,\n",
    "                             xlabel_rotation=0): \n",
    "    import seaborn as sns\n",
    "    fig, ax = plt.subplots(figsize=(8,6))\n",
    "    fig.patch.set_facecolor(\"white\")\n",
    "    \n",
    "    cmap = None\n",
    "    if cluster_palette is not None:\n",
    "        cmap = sns.palettes.blend_palette(\n",
    "            cluster_palette, \n",
    "            n_colors=len(cluster_palette), \n",
    "            as_cmap=True)\n",
    "   \n",
    "    cluster_props.plot(\n",
    "        kind=\"bar\", \n",
    "#         stacked=True, \n",
    "        ax=ax, \n",
    "        legend=None, \n",
    "        colormap=cmap\n",
    "    )\n",
    "    \n",
    "    fontsize=16\n",
    "    ax.legend(bbox_to_anchor=(1.1, 1.2), frameon=False, title=\"Assay type\", fontsize=fontsize)\n",
    "    sns.despine(fig, ax)\n",
    "    ax.tick_params(axis=\"x\", rotation=xlabel_rotation)\n",
    "    ax.set_xlabel(cluster_props.index.name.capitalize(), fontsize=fontsize)\n",
    "    ax.set_ylabel(\"% of nuclei in cluster\")\n",
    "    ax.set_xticklabels(cluster_props.index, rotation = 90, fontsize=fontsize)\n",
    "    ax.set_yticks([0,50,100])\n",
    "    ax.set_yticklabels([0,50,100], fontsize=fontsize)\n",
    "    \n",
    "    fig.tight_layout()\n",
    "    \n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def jaccard(list1, list2):\n",
    "    intersection = len(list(set(list1).intersection(list2)))\n",
    "    union = (len(list1) + len(list2)) - intersection\n",
    "    return float(intersection) / union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_jaccard_pathways(pathways, pathway_dict): \n",
    "    import pandas as pd\n",
    "    \n",
    "    path_ij = np.zeros((len(pathways),len(pathways)))\n",
    "    values = []\n",
    "\n",
    "    for i in range(len(pathways)):\n",
    "        p1 = pathways[i]\n",
    "        for j in range(i+1,len(pathways)):\n",
    "            p2 = pathways[j]\n",
    "            J = jaccard(pathway_dict[p1],pathway_dict[p2])\n",
    "            path_ij[i,j] = J\n",
    "            path_ij[j,i] = J\n",
    "            values.append(J)\n",
    "\n",
    "    path_ij = pd.DataFrame(path_ij, columns=pathways, index=pathways)\n",
    "    \n",
    "    return path_ij, values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_path_Jij(df_J): \n",
    "    \n",
    "    import random\n",
    "    random.seed(0)\n",
    "    \n",
    "    plt.figure(figsize=(20,20))\n",
    "    sns.set(font_scale=0.2)\n",
    "    cg = sns.clustermap(df_J,\n",
    "                        yticklabels=True,\n",
    "                        xticklabels=False,\n",
    "                        vmin=0,\n",
    "                        vmax=1,\n",
    "                        method='ward',\n",
    "                        cmap='Blues')\n",
    "    cg.ax_col_dendrogram.set_visible(False)\n",
    "    plt.savefig('%s/pathway_dendrogram_ward.pdf' %(sc.settings.figdir),\n",
    "                   bbox_inches='tight')\n",
    "    \n",
    "    return cg\n",
    "\n",
    "def extract_cluster_feats(cg):\n",
    "    import random\n",
    "    random.seed(0)\n",
    "    \n",
    "    # reordered pathway idx\n",
    "    ordered_idx = cg.dendrogram_col.reordered_ind\n",
    "#     reordered_paths = [pathways[i] for i in ordered_idx]\n",
    "\n",
    "    # extract tree linkage from clustergram\n",
    "    Z = cg.dendrogram_col.linkage\n",
    "    return ordered_idx, Z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cut_tree(Z, pathways, CUT_THRESHOLD=2.3): \n",
    "    import random\n",
    "    random.seed(0)\n",
    "    \n",
    "    from scipy import cluster\n",
    "    from scipy.cluster.hierarchy import fcluster, fclusterdata, dendrogram\n",
    "    import pandas as pd\n",
    "    \n",
    "    # plot tree cutting threshold\n",
    "    sns.reset_orig()\n",
    "    plt.figure(figsize=(8,15))\n",
    "    dn2 = dendrogram(Z, orientation='left', no_labels=True,\n",
    "                     color_threshold=CUT_THRESHOLD,\n",
    "                    above_threshold_color='#bcbcbc')\n",
    "    plt.axvline(CUT_THRESHOLD, color='k', linestyle='--')\n",
    "    plt.gca().invert_yaxis()\n",
    "    plt.xlabel('Dendrogram distance')\n",
    "    plt.savefig('%s/pathway_dendrogram_tree_cut_threshold.pdf' %(sc.settings.figdir),\n",
    "                   bbox_inches='tight')\n",
    "\n",
    "    assignments = fcluster(Z, CUT_THRESHOLD, criterion='distance')\n",
    "    cluster_output = pd.DataFrame({'pathway':pathways, 'cluster':assignments})\n",
    "    cluster_output.set_index('pathway', inplace=True)\n",
    "    \n",
    "    return cluster_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pylab import *\n",
    "\n",
    "def make_pathway_cmap(cluster_assign, pathways): \n",
    "    cmap = cm.get_cmap('Set2', len(set(cluster_assign['cluster_reduced'])))\n",
    "    colors = []\n",
    "    for i in range(cmap.N):\n",
    "        rgba = cmap(i)\n",
    "        colors.append(matplotlib.colors.rgb2hex(rgba))\n",
    "\n",
    "    np.random.shuffle(colors)\n",
    "    color_dict = dict(zip(set(cluster_assign['cluster_reduced']), colors))\n",
    "    color_dict[0] = '#ffffff'\n",
    "    \n",
    "    \n",
    "    df_colors = pd.DataFrame(0, index=pathways, columns=['color'])\n",
    "    for f, path in zip(cluster_assign['cluster_reduced'], pathways):\n",
    "        df_colors.loc[path] = color_dict[f]\n",
    "        \n",
    "    return df_colors, color_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_GTEx_lung_data(): \n",
    "    import scanpy as sc\n",
    "    import pandas as pd\n",
    "    \n",
    "    ad_gtex = sc.read('GTEx_data/GTEx_8_tissues_snRNAseq_atlas_071421.public_obs.h5ad')\n",
    "    ad_gtex_lung = ad_gtex[ad_gtex.obs['tissue']=='lung'].copy()\n",
    "    \n",
    "    ad_gtex_lung.layers['scaled'] = ad_gtex_lung.X.copy()\n",
    "    sc.pp.scale(ad_gtex_lung, layer='scaled', max_value=10)\n",
    "    \n",
    "    return ad_gtex_lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_LUAD_data(): \n",
    "    import scanpy as sc\n",
    "    import pandas as pd\n",
    "\n",
    "    ad_LUAD = sc.read('GSE131907_raw.h5ad')\n",
    "    LUAD_metadata = pd.read_csv('GSE131907_Lung_Cancer_cell_annotation.txt', delimiter='\\t')\n",
    "    LUAD_metadata.set_index('Index', inplace=True)\n",
    "    ad_LUAD.obs = pd.concat([ad_LUAD.obs, LUAD_metadata], axis=1, join=\"inner\")\n",
    "    \n",
    "    ad_LUAD.obs['n_counts'] = ad_LUAD.obs['nCount_RNA'].copy()\n",
    "    ad_LUAD.obs['n_genes'] = ad_LUAD.obs['nFeature_RNA'].copy()\n",
    "    ad_LUAD.obs['Sample ID'] = ad_LUAD.obs['Sample'].copy()\n",
    "    \n",
    "    ad_LUAD_lung = ad_LUAD[ad_LUAD.obs['Sample_Origin'].isin(['tLung','nLung']),:].copy()\n",
    "    ad_LUAD_lung.layers['counts'] = ad_LUAD_lung.X.copy()\n",
    "    \n",
    "    # match adata.obs['\"Cell types level 2\"']\n",
    "    broad_class_dict = {'B lymphocytes':'Immune (lymphocyte)',\n",
    "                         'Endothelial cells':'Endothelial cell',\n",
    "                         'Epithelial cells':'Epithelial cell',\n",
    "                         'Fibroblasts':'Fibroblast',\n",
    "                         'MAST cells':'Immune (myeloid)',\n",
    "                         'Myeloid cells':'Immune (myeloid)',\n",
    "                         'T/NK cells':'Immune (lymphocyte)'\n",
    "                       }\n",
    "    ad_LUAD_lung.obs['Cell types level 2'] = ad_LUAD_lung.obs['Cell_type.refined'].map(broad_class_dict)\n",
    "    \n",
    "    # match 'Broad cell type (numbers)'\n",
    "    cell_type_dict = {'AT1':'5. Epithelial cell (alveolar type I)',\n",
    "                    'AT2':'6. Epithelial cell (alveolar type II)',\n",
    "                    'Activated DCs':'21. Immune (DC/macrophage)',\n",
    "                    'Alveolar Mac':'25. Immune (alveolar macrophage)',\n",
    "                    'CD141+ DCs':'21. Immune (DC/macrophage)',\n",
    "                    'CD163+CD14+ DCs':'21. Immune (DC/macrophage)',\n",
    "                    'CD1c+ DCs':'21. Immune (DC/macrophage)',\n",
    "                    'CD207+CD1a+ LCs':'21. Immune (DC/macrophage)',\n",
    "                    'CD4+ Th':'24. Immune (T cell)',\n",
    "                    'CD8 low T':'24. Immune (T cell)',\n",
    "                    'CD8+/CD4+ Mixed Th':'24. Immune (T cell)',\n",
    "                    'COL13A1+ matrix FBs':'17. Fibroblast',\n",
    "                    'COL14A1+ matrix FBs':'17. Fibroblast',\n",
    "                    'Ciliated':'9. Epithelial cell (ciliated)',\n",
    "                    'Club':'10. Epithelial cell (club)',\n",
    "                    'Cytotoxic CD8+ T':'24. Immune (T cell)',\n",
    "                    'EPCs':'3. Endothelial cell (vascular)',\n",
    "                    'Exhausted CD8+ T':'24. Immune (T cell)',\n",
    "                    'Exhausted Tfh':'24. Immune (T cell)',\n",
    "                    'FB-like cells':'17. Fibroblast',\n",
    "                    'Follicular B cells':'19. Immune (B cell)',\n",
    "                    'GC B cells in the DZ':'19. Immune (B cell)',\n",
    "                    'GC B cells in the LZ':'19. Immune (B cell)',\n",
    "                    'GrB-secreting B cells':'19. Immune (B cell)',\n",
    "                    'Lymphatic ECs':'2. Endothelial cell (lymphatic)',\n",
    "                    'MALT B cells':'19. Immune (B cell)',\n",
    "                    'MAST':'26. Immune (mast cell)',\n",
    "                    'Malignant cells':'44. Malignant cells', # new\n",
    "                    'Mesothelial cells':'45. Mesothelial cells', # new\n",
    "                    'Microglia/Mac':'21. Immune (DC/macrophage)',\n",
    "                    'Monocytes':'21. Immune (DC/macrophage)',\n",
    "                    'Myofibroblasts':'17. Fibroblast',\n",
    "                    'NK':'23. Immune (NK cell)',\n",
    "                    'Naive CD4+ T':'24. Immune (T cell)',\n",
    "                    'Naive CD8+ T':'24. Immune (T cell)',\n",
    "                    'Pericytes':'39. Pericyte/SMC',\n",
    "                    'Plasma cells':'19. Immune (B cell)',\n",
    "                    'Pleural Mac':'21. Immune (DC/macrophage)',\n",
    "                    'Smooth muscle cells':'39. Pericyte/SMC',\n",
    "                    'Stalk-like ECs':'3. Endothelial cell (vascular)',\n",
    "                    'Tip-like ECs':'3. Endothelial cell (vascular)',\n",
    "                    'Treg':'24. Immune (T cell)',\n",
    "                    'Tumor ECs':'40. Endothelial cell (tumor)', # new\n",
    "                    'mo-Mac':'21. Immune (DC/macrophage)',\n",
    "                    'pDCs':'21. Immune (DC/macrophage)',\n",
    "                    'tS1':'41. tumor State 1', # new\n",
    "                    'tS2':'42. tumor State 2', # new\n",
    "                    'tS3':'43. tumor State 3' # new\n",
    "                 } \n",
    "    ad_LUAD_lung.obs['Cell type (numbers)'] = ad_LUAD_lung.obs['Cell_subtype'].map(cell_type_dict)\n",
    "    \n",
    "    ad_LUAD_lung.obs.drop(columns=['nCount_RNA','nFeature_RNA', 'Barcode', 'orig.ident'], inplace=True)\n",
    "    \n",
    "    ad_LUAD_lung.layers['counts'] = ad_LUAD_lung.X.copy()\n",
    "    sc.pp.normalize_total(ad_LUAD_lung)\n",
    "    sc.pp.log1p(ad_LUAD_lung)\n",
    "    sc.pp.scale(ad_LUAD_lung, max_value=10)\n",
    "    ad_LUAD_lung.layers['scaled'] = ad_LUAD_lung.X.copy()\n",
    "    \n",
    "    return ad_LUAD_lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cell_type_markers(ad, group_name, num_genes=50, test_method='t-test', use_raw=True): \n",
    "    if group_name == 'GTEx':\n",
    "        group_name = 'Cell types level 2'\n",
    "    elif group_name == 'LUAD_scRNA':\n",
    "        group_name = 'Cell_type'\n",
    "        \n",
    "    sc.tl.rank_genes_groups(ad, group_name, method=test_method, use_raw=use_raw)\n",
    "    cluster_markers = pd.DataFrame(ad.uns['rank_genes_groups']['names']).head(num_genes)\n",
    "    marker_dict = cluster_markers.to_dict(orient='list')\n",
    "    \n",
    "    return marker_dict, cluster_markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def valid_markers(marker_dict, ad_target): \n",
    "    marker_dict_valid = {}\n",
    "\n",
    "    for k,v in marker_dict.items(): \n",
    "        valid_genes = [g for g in v if g in ad_target.var_names]\n",
    "        print(k, len(valid_genes))\n",
    "        marker_dict_valid[k] = valid_genes\n",
    "        \n",
    "    return marker_dict_valid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def score_with_markers(marker_dict_valid, ad_score, target_clusters, score_base): \n",
    "    for clust in target_clusters: \n",
    "        sc.tl.score_genes(ad_score, marker_dict_valid[clust], score_name='%s_%s' %(score_base, clust))\n",
    "\n",
    "    score_names = [score_base+'_'+clust for clust in target_clusters]\n",
    "    \n",
    "    for i in range(len(score_names)):\n",
    "        plt.figure()\n",
    "        ad_score.obs[score_names[i]].hist(bins=50, alpha=0.6)\n",
    "        plt.xlabel(score_names[i])\n",
    "        plt.axvline(0, color='k')\n",
    "        plt.grid(False)\n",
    "        plt.savefig('%s/cell_type_score_%s_%s.pdf' %(sc.settings.figdir, score_base, score_names[i]))\n",
    "    \n",
    "    ad_score.obs['rank_assignment'] = ad_score.obs[score_names].rank(ascending=False).idxmin(axis=1)\n",
    "    ad_score.obs['max_score_assignment'] = ad_score.obs[score_names].idxmax(axis=1)\n",
    "    \n",
    "    return ad_score, score_names\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
